File: app/src/main/java/io/github/ichisadashioko/android/kanji/tflite/KanjiClassifier.java
package io.github.ichisadashioko.android.kanji.tflite;

import android.app.Activity;
import android.content.res.AssetFileDescriptor;
import android.graphics.Bitmap;
import android.util.Log;

import org.tensorflow.lite.Interpreter;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.util.*;

public class KanjiClassifier
{
    public static final int MAX_RESULTS    = 16;
    public static final int DIM_BATCH_SIZE = 1;
    public static final int DIM_PIXEL_SIZE = 1;
    public static final String LOG_TAG     = "KanjiClassifier";
    public static final int IMAGE_WIDTH    = 64;
    public static final int IMAGE_HEIGHT   = 64;
    // we will use 32-bit float to store pixel value
    public static final int NUM_BYTES_PER_PIXEL = 4;
    public static final String MODEL_FILE_PATH  = "etlcb_9b_model.tflite";
    public static final String LABEL_FILE_PATH  = "etlcb_9b_labels.txt";

    // pre-allocated buffers to store image data
    public int[] intValues = new int[IMAGE_WIDTH * IMAGE_HEIGHT];
    public final int NUM_LABELS;
    public Interpreter.Options tfliteOptions;
    public MappedByteBuffer tfliteModel;
    public List<String> labels;
    public Interpreter tflite;
    public ByteBuffer imgData;
    public float[][] labelProbArray;

    public KanjiClassifier(Activity activity) throws IOException
    {
        tfliteModel   = loadModelFile(activity);
        tfliteOptions = new Interpreter.Options();
        tfliteOptions.setNumThreads(Runtime.getRuntime().availableProcessors());
        tfliteOptions.setUseNNAPI(true);
        tflite = new Interpreter(tfliteModel, tfliteOptions);

        labels     = loadLabelList(activity);
        NUM_LABELS = labels.size();

        imgData = ByteBuffer.allocateDirect(DIM_BATCH_SIZE * IMAGE_HEIGHT * IMAGE_WIDTH * DIM_PIXEL_SIZE * NUM_BYTES_PER_PIXEL);
        imgData.order(ByteOrder.nativeOrder());
        labelProbArray = new float[DIM_BATCH_SIZE][NUM_LABELS];

        Log.d(LOG_TAG, "Created Kanji Classifier.");
    }

    public MappedByteBuffer loadModelFile(Activity activity) throws IOException
    {
        AssetFileDescriptor afd = activity.getAssets().openFd(MODEL_FILE_PATH);
        FileInputStream fis     = new FileInputStream(afd.getFileDescriptor());
        FileChannel fileChannel = fis.getChannel();
        return fileChannel.map(FileChannel.MapMode.READ_ONLY, afd.getStartOffset(), afd.getDeclaredLength());
    }

    public List<String> loadLabelList(Activity activity) throws IOException
    {
        ArrayList<String> labels = new ArrayList<>();
        BufferedReader reader    = new BufferedReader(new InputStreamReader(activity.getAssets().open(LABEL_FILE_PATH)));
        String line              = reader.readLine();
        while (line != null)
        {
            labels.add(line);
            line = reader.readLine();
        }
        return labels;
    }

    public float normalizePixelValue(int pixelValue)
    {
        int r      = (pixelValue >> 16) & 0xFF;
        int g      = (pixelValue >> 8) & 0xFF;
        int b      = pixelValue & 0xFF;
        float gray = (0.299f * r + 0.597f * g + 0.114f * b) / 255f;
        return gray;
    }

    public void populateByteBuffer(Bitmap bitmap) throws Exception
    {
        if ((bitmap.getWidth() != IMAGE_WIDTH) && (bitmap.getHeight() != IMAGE_HEIGHT))
        {
            throw new Exception(String.format("The image with shape (%d, %d) is not equals (%d, %d)!!!", bitmap.getWidth(), bitmap.getHeight(), IMAGE_WIDTH, IMAGE_HEIGHT));
        }
        // reset `imgData`
        imgData.rewind();
        // populate `intValues` with the bitmap data
        bitmap.getPixels(intValues, 0, IMAGE_WIDTH, 0, 0, IMAGE_WIDTH, IMAGE_HEIGHT);
        int index = 0;
        for (int i = 0; i < IMAGE_WIDTH; i++)
        {
            for (int j = 0; j < IMAGE_HEIGHT; j++)
            {
                int pixelValue = intValues[index++];
                imgData.putFloat(normalizePixelValue(pixelValue));
            }
        }
    }

    public synchronized List<Recognition> recognizeImage(Bitmap bitmap)
    {
        ArrayList<Recognition> results = new ArrayList<>();

        try
        {
            populateByteBuffer(bitmap);
        }
        catch (Exception ex)
        {
            Log.e(LOG_TAG, "There is some problem with the Bitmap!");
            ex.printStackTrace();
            return results;
        }

        tflite.run(imgData, labelProbArray);
        long timestamp = System.currentTimeMillis();

        // sort the result by confidence
        PriorityQueue<Recognition> pq = new PriorityQueue<>(NUM_LABELS, new Comparator<Recognition>() {
            @Override
            public int compare(Recognition a, Recognition b)
            {
                // we want to sort descending
                return Float.compare(b.confidence, a.confidence);
            }
        });

        for (int i = 0; i < NUM_LABELS; i++)
        {
            pq.add(new Recognition(i, timestamp, labels.get(i), labelProbArray[0][i]));
        }

        int returnSize = Math.min(pq.size(), MAX_RESULTS);
        for (int i = 0; i < returnSize; i++)
        {
            results.add(pq.poll());
        }

        return results;
    }
}

File: app/src/main/java/com/example/kanji_recognition_tflite/KanjiClassifier.kt
package com.example.kotlindrawing

import android.app.Activity
import android.graphics.Bitmap
import android.os.Build
import android.os.SystemClock
import android.os.Trace
import android.support.annotation.RequiresApi
import android.util.Log
import org.tensorflow.lite.Interpreter
import java.io.BufferedReader
import java.io.FileInputStream
import java.io.InputStreamReader
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import java.util.*
import kotlin.collections.ArrayList

/**
 * A classifier specialized to label images using TensorFlow Lite
 */
class KanjiClassifier constructor(activity: Activity) {
    companion object {
        // static objects are put here.

        // Number of results to show in the UI.
        const val MAX_RESULTS = 128
        // Dimensions of inputs.
        const val DIM_BATCH_SIZE = 1
        const val DIM_PIXEL_SIZE = 1
        const val LOG_TAG = "KanjiClassifier"
    }

    /**
     * Preallocated buffers to store image data in.
     */
    var intValues = IntArray(getImageSizeX() * getImageSizeY())
    /**
     * Options for configuring the Interpreter.
     */
    private var tfliteOptions: Interpreter.Options
    /**
     * The loaded Tensorflow Lite model.
     */
    private var tfliteModel: MappedByteBuffer?
    /**
     * Labels corresponding to the output of the model
     */
    private var labels: List<String>
    /**
     * An instance of the driver class to run model inference with Tensorflow Lite.
     */
    private var tflite: Interpreter?
    /**
     * The number of threads to use for classification.
     */
    var numThreads = 4
    private var imgData: ByteBuffer

    /**
     * An array to hold inference results, to be fed into Tensorflow Lite as outputs.
     */
    private var labelProbArray: Array<FloatArray>

    /**
     * Get the total number of labels.
     */
    val numLabels: Int
        get() = labels.size

    init {
        // initial Tensorflow Lite Interpreter
        tfliteModel = loadModelFile(activity)
        tfliteOptions = Interpreter.Options()
        tfliteOptions.setNumThreads(numThreads)
        tfliteOptions.setUseNNAPI(true)
        tflite = Interpreter(tfliteModel!!, tfliteOptions)
        labels = loadLabelList(activity)

        imgData = ByteBuffer.allocateDirect(
            DIM_BATCH_SIZE
                    * getImageSizeY()
                    * getImageSizeX()
                    * DIM_PIXEL_SIZE
                    * getNumBytesPerChannel()
        )
        imgData.order(ByteOrder.nativeOrder())
        Log.d(LOG_TAG, "Created Kanji Classifier")

        labelProbArray = Array(1) { FloatArray(numLabels) }
    }

    fun getNumBytesPerChannel(): Int {
        // Float.SIZE / Byte.SIZE
        // return java.lang.Float.SIZE
        return 4 // java.nio.ByteBuffer.putFloat() only writes 4 bytes to the buffer
    }

    fun getImageSizeX(): Int {
        return 64
    }

    fun getImageSizeY(): Int {
        return 64
    }

    fun getModelPath(): String {
        // model filename in assets folder
        return "etlcb_9b_model.tflite"
    }

    fun getLabelPath(): String {
        // label filename in assets folder
        return "etlcb_9b_labels.txt"
    }

    /**
     * Memory-map the model file in Assets.
     */
    private fun loadModelFile(activity: Activity): MappedByteBuffer {
        val fileDescriptor = activity.assets.openFd(getModelPath())
        val inputStream = FileInputStream(fileDescriptor.fileDescriptor)

        val fileChannel = inputStream.channel
        var startOffset = fileDescriptor.startOffset
        var declaredLength = fileDescriptor.declaredLength
        return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
    }

    /**
     * Reads label list from Assets.
     */
    private fun loadLabelList(activity: Activity): List<String> {
        val labels = ArrayList<String>()
        val reader = BufferedReader(InputStreamReader(activity.assets.open(getLabelPath())))
        var line = reader.readLine()
        while (line != null) {
            labels.add(line)
            line = reader.readLine()
        }
        return labels
    }

    /**
     * Writes Image data into a `ByteBuffer`.
     */
    private fun convertBitmapToByteBuffer(bitmap: Bitmap) {
        imgData.rewind()
        bitmap.getPixels(intValues, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
        // Convert the image to floating point.
        var pixel = 0
        val startTime = SystemClock.uptimeMillis()
        for (i in 0 until getImageSizeX()) {
            for (j in 0 until getImageSizeY()) {
                val pxVal = intValues[pixel++]
                addPixelValue(pxVal)
            }
        }
        val endTime = SystemClock.uptimeMillis() - startTime
        Log.v(LOG_TAG, "Time cost to put values into ByteBuffer: $endTime")
    }

    /**
     * Add pixelValue to byteBuffer.
     *
     * The `pixelValue` from `Bitmap` is a 32-bits integer for ARGB.
     *
     */
    protected fun addPixelValue(pixelValue: Int) {
        // Convert RGB to GRAY
        val red = pixelValue shr 16 and 0xff
        val green = pixelValue shr 8 and 0xff
        val blue = pixelValue and 0xff
        val gray = (0.299f * red + 0.597f * green + 0.114f * blue) / 255f
        imgData.putFloat(gray)

        // Use the below code to put RGB values
//        imgData.putFloat((pixelValue shr 16 and 0xFF) / 255f)
//        imgData.putFloat((pixelValue shr 8 and 0xFF) / 255f)
//        imgData.putFloat((pixelValue and 0xFF) / 255f)
    }

    /**
     * Runs inference and returns the classification results.
     */
    @RequiresApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
    fun recognizeImage(bitmap: Bitmap): List<Recognition> {
        // Log this method so that it can be analyzed with systrace.
        Trace.beginSection("recognizeImage")

        Trace.beginSection("preprocessBitmap")
        convertBitmapToByteBuffer(bitmap)
        Trace.endSection()

        // Run the inference call.
        Trace.beginSection("runInference")
        val startTime = SystemClock.uptimeMillis()
        runInference()
        val endTime = SystemClock.uptimeMillis() - startTime
        Trace.endSection()
        Log.v(LOG_TAG, "Time cost to run model inference: $endTime")

        // Find the best classification.
        val pq = PriorityQueue(
            3,
            Comparator<Recognition> { lhs, rhs ->
                // Intentionally reversed to put high confidence at the head of the queue.
                java.lang.Float.compare(rhs.confidence, lhs.confidence)
            }
        )
        for (i in labels.indices) {
            pq.add(
                Recognition(
                    "" + i,
                    if (labels.size > i) labels[i] else "unknown",
                    getNormalizedProbability(i)
                )
            )
        }
        val recognitions = ArrayList<Recognition>()
        val recognitionsSize = Math.min(pq.size, MAX_RESULTS)
        for (i in 0 until recognitionsSize) {
            recognitions.add(pq.poll())
        }
        Trace.endSection()
        return recognitions
    }

    /**
     * Get the normalized probability value for the specified label. This is the final value as it will be shown to the user.
     */
    fun getNormalizedProbability(labelIndex: Int): Float {
        return labelProbArray[0][labelIndex]
    }

    /**
     * Run inference using the prepared input in `imgData`. Afterwards, the result will be provided by `getProbability()`.
     *
     * This additional method is necessary, because we don't have a common base for different primitive data types.
     */
    fun runInference() {
        tflite!!.run(imgData, labelProbArray)
    }

    /**
     * Closes the intepreter and model to release resources.
     */
    fun close() {
        tflite!!.close()
        tflite = null
        tfliteModel = null
    }

    /**
     * An immutable result returned by a Classifier describing what was recognized.
     * @param id A unique identifier for what has been recognized. Specific to the class, not the instance of the object
     * @param title Display name for the recognition.
     * @param confidence A sortable score for how good the recognition is relative to others. Higher should be better.
     */
    class Recognition constructor(_id: String?, _title: String?, _confidence: Float) {
        val id = _id
        val title = _title
        val confidence = _confidence

        override fun toString(): String {
            var resultString = "";
            if (id != null) {
                resultString += String.format("%-8s", "[$id]")
            }
            if (title != null) {
                resultString += String.format("%-5s", "$title")
            }
            resultString += String.format("(%.1f%%) ", confidence * 100f)

            return resultString;
        }
    }
}